packages/components/src/public/files/TemplateDownloadButton.tsx

```diff
@@ -1,35 +1,133 @@
-import React, { FC, memo } from 'react';
+import React, { FC, memo, useCallback, useEffect, useMemo, useState } from 'react';
 
 import { PermissionTypes } from '@labkey/api';
 
 import { User } from '../../internal/components/base/models/User';
 import { RequiresPermission } from '../../internal/components/base/Permissions';
+import { DropdownButton, MenuItem } from '../../internal/dropdowns';
+import { ImportTemplate } from '../QueryInfo';
+import { LoadingSpinner } from '../../internal/components/base/LoadingSpinner';
+import { SchemaQuery } from '../SchemaQuery';
+import { useAppContext } from '../../internal/AppContext';
 
 interface Props {
     className?: string;
-    onClick?: () => void;
-    templateUrl?: string;
+    defaultTemplateUrl?: string;
+    getExtraTemplates?: () => Promise<ImportTemplate[]>;
+    onDownloadDefault?: () => void;
+    schemaQuery?: SchemaQuery;
     text?: string;
     user?: User;
 }
 
 export const TemplateDownloadButton: FC<Props> = memo(props => {
-    const { className, onClick, templateUrl, text = 'Template', user } = props;
+    const {
+        schemaQuery,
+        getExtraTemplates,
+        className,
+        onDownloadDefault,
+        defaultTemplateUrl,
+        text = 'Template',
+        user,
+    } = props;
+    const [customTemplates, setCustomTemplates] = useState<ImportTemplate[]>();
+    const [loadingTemplates, setLoadingTemplates] = useState<boolean>(false);
+    const { api } = useAppContext();
 
-    if (!onClick && !templateUrl?.length) return null;
+    useEffect(() => {
+        if (!schemaQuery) return;
+
+        (async () => {
+            try {
+                setLoadingTemplates(true);
+                const queryInfo = await api.query.getQueryDetails({
+                    schemaName: schemaQuery.schemaName,
+                    queryName: schemaQuery.queryName,
+                });
+                setCustomTemplates(queryInfo.getCustomTemplates());
+            } catch (reason) {
+                console.error(reason);
+            } finally {
+                setLoadingTemplates(false);
+            }
+        })();
+    }, [schemaQuery, setLoadingTemplates]);
+
+    const showDropdown = useMemo(() => {
+        return customTemplates?.length > 0 || (!!getExtraTemplates && !customTemplates);
+    }, [customTemplates, getExtraTemplates]);
+
+    const fetchTemplates = useCallback(async () => {
```

Comment:
as discussed, this will change when `getExtraTemplates` is removed

Reply:
done